File: src/2023/day-02/typescript/part-2.ts
export function Part2(input: string) {
  return input
    .trim()
    .split("\n")
    .map((line) => {
      const [_, game] = line.split(": ");
      const max = { r: 0, g: 0, b: 0 }
      game.split("; ").map((set) =>
        set.split(", ").map((score) => {
          const [v, c] = score.split(" ")
          if (c === "red" && +v > max.r) max.r = +v
          else if (c === "green" && +v > max.g) max.g = +v
          else if (c === "blue" && +v > max.b) max.b = +v
        })
      )
      return max.r * max.g * max.b;
    }).reduce((a, b) => a + b);
}

File: src/2023/day-02/rust/src/part_2.rs
use itertools::Itertools;

pub fn part_2(input: &str) -> i32 {
  input
    .trim()
    .split('\n')
    .map(|line| {
      let (_, games) = line.split(": ").collect_tuple().unwrap();
      let mut max = (0, 0, 0); // red, green, blue

      games.split("; ").for_each(|set| {
        set.split(", ").for_each(|score| {
          let (v, color) = score.split(' ').collect_tuple().unwrap();
          let v: i32 = v.parse().unwrap();

          match color {
            "red" => max.0 = max.0.max(v),
            "green" => max.1 = max.1.max(v),
            "blue" => max.2 = max.2.max(v),
            _ => (),
          }
        });
      });

      max.0 * max.1 * max.2
    })
    .sum()
}
